google/cloud/security/scanner/audit/org_rules_engine.py

```diff
@@ -31,6 +31,81 @@
 from google.cloud.security.scanner.audit.errors import InvalidRulesSchemaError
 
 
+def _check_whitelist_members(rule_members=None, policy_members=None):
+    """Whitelist: Check that policy members ARE in rule members.
+
+    If a policy member is NOT found in the rule members, add it to
+    the violating members.
+
+    Args:
+        rule_members: A list of IamPolicyMembers allowed in the rule.
+        policy_members: A list of IamPolicyMembers in the policy.
+
+    Return:
+        A list of the violating members: policy members NOT found in
+        the whitelist (rule members).
+    """
+    violating_members = []
+    for policy_member in policy_members:
+        is_whitelisted = False
+        for rule_member in rule_members:
+            is_whitelisted = rule_member.matches(policy_member)
+            if is_whitelisted:
+                break
+        if not is_whitelisted:
+            violating_members.append(policy_member)
+    return violating_members
+
+def _check_blacklist_members(rule_members=None, policy_members=None):
+    """Blacklist: Check that policy members ARE NOT in rule members.
+
+    If a policy member is found in the rule members, add it to the
+    violating members.
+
+    Args:
+        rule_members: A list of IamPolicyMembers allowed in the rule.
+        policy_members: A list of IamPolicyMembers in the policy.
+
+    Return:
+        A list of the violating members: policy members found in
+        the blacklist (rule members).
+    """
+    violating_members = [
+        policy_member
+        for policy_member in policy_members
+        for rule_member in rule_members
+        if rule_member.matches(policy_member)
+    ]
+    return violating_members
+
+def _check_required_members(rule_members=None, policy_members=None):
+    """Required: Check that rule members are in policy members.
+
+    If a required rule member is NOT found in the policy members, add
+    it to the violating members. Note that the check is different:
+    it's reversed from the whitelist/blacklist (policy as a subset of
+    rules vs rules as subset of policy).
+
+    Args:
+        rule_members: A list of IamPolicyMembers allowed in the rule.
+        policy_members: A list of IamPolicyMembers in the policy.
+
+    Return:
+        A list of the violating members: rule members not found in the
+        policy (required-whitelist).
+    """
+    violating_members = []
+    for rule_member in rule_members:
+        found_required_member = False
+        for policy_member in policy_members:
```

Comment:
Maybe change to:

found_required_member = any(rule_member.matches(m) for m in policy_members)